src/Chainweb/BlockHeaderDB/PruneForks.hs

```diff
@@ -61,160 +99,162 @@ import Data.LogMessage
-              else
-                -- The key is either
-                -- 1. in the chain or
-                -- 2. is a false positive that has a payload and is connected to
-                --    the chain (i.e. has all of its predecessors).
-                --
-                -- We accept a small number of nodes of the second case.
-                --
-                return (falsePositiveSet, i)
-
-    deleteKey h isPayloadMarked = do
+    go :: ([BlockHash], Int) -> [BlockHeader] -> IO ([BlockHash], Int)
+    go ([], _) _ = throwM $ InternalInvariantViolation "PrunForks.pruneForks_: impossible case"
+    go (!pivots, !n) curs = do
+
+        let (newPivots, toDelete) = L.partition (\h -> _blockHash h `elem` pivots) curs
+            pivotPayloads = _blockPayloadHash <$> newPivots
+
+        -- TODO: try to repair the database by fetching the missing block from
+        -- remote peers?
+        --
+        -- It's probably the best to write an independent repair program or
+        -- module
+        when (null newPivots) $ do
+            logg Error
+                $ "Corrupted chain database for chain " <> toText (_chainId cdb)
+                <> ". The chain db must be deleted and re-resynchronized."
+            throwM $ TreeDbKeyNotFound @BlockHeaderDb (head pivots)
+            -- note that the use of `head` here is safe
+
+        forM_ toDelete $ \b -> do
+            deleteHdr b
+            hdrCallback True b
+            let payload = _blockPayloadHash b
+            payloadCallback (payload `notElem` pivotPayloads) payload
+
+        forM_ newPivots $ \b -> do
+            hdrCallback False b
+            payloadCallback False (_blockPayloadHash b)
+
+        return (_blockParent <$> newPivots, n + length toDelete)
+
+
+    deleteHdr k = do
         -- TODO: make this atomic (create boilerplate to combine queries for
         -- different tables)
-        casDelete (_chainDbCas cdb) (casKey $ RankedBlockHeader h)
-        tableDelete (_chainDbRankTable cdb) (_blockHash h)
-        logg Debug $ "deleted block header at height " <> sshow (_blockHeight h) <> " with payload mark " <> sshow isPayloadMarked
-        callback h isPayloadMarked
+        casDelete (_chainDbCas cdb) (casKey $ RankedBlockHeader k)
+        tableDelete (_chainDbRankTable cdb) (_blockHash k)
+        logg Debug
+            $ "pruned block header " <> encodeToText (_blockHash k)
+            <> " at height " <> sshow (_blockHeight k)
+
+-- -------------------------------------------------------------------------- --
+-- Utils
+
+-- TODO: provide this function in chainweb-storage:
+--
+-- Returns the stream of key-value pairs of an 'RocksDbTableIter' in reverse
+-- order.
+--
+-- The iterator must be released after the stream is consumed. Releasing the
+-- iterator to early while the stream is still in use results in a runtime
+-- error. Not releasing the iterator after the processing of the stream has
+-- finished results in a memory leak.
+--
```

Comment:
What is the use case for exposing this outside of `withReverseHeaderStream` when it seems like the latter is the preferred function for safe usage?

Reply:
Copy and paste from the respective non-reversed functions in chainweb-storage :-)